### PR TITLE
513 - Add DDF data augmentation

### LINKED ISSUE #513: Apply DDF transformation as data augmentation
## Subject of the feature

What are you trying to do and how would you want to do it differently?
From issue #424 and @acasamitjana 
"Data augmentation: apart from the existing AffineTF, we can include DDF transformations as well"

Is it something you currently you cannot do?
Unsure.

Is this related to an issue/problem?
No.

Has the feature been requested before? If yes, please provide a link to the issue.
No.

If the feature request is approved, would you be willing to submit a PR?

M.

### PR DESCRIPTION
**Pull request by acasamitjana**
# Description

Add optional spatial transformation as data augmentation, with strength and smoothness (lowres_size) as input parameters.

Fixes #513 

## Type of change

What types of changes does your code introduce to DeepReg? _Put an `x` in the boxes that
apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [X] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [X] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by acasamitjana** [comment]
Here's the WIP PR @YipengHu @mathpluscode. Can you take a look at what's currently implemented before I move to define new test for all new functionalities?

File: deepreg/dataset/loader/interface.py [hunk 1]
```diff
@@ -120,16 +125,32 @@ def get_dataset_and_preprocess(
         dataset = dataset.batch(batch_size=batch_size, drop_remainder=training)
         dataset = dataset.prefetch(tf.data.experimental.AUTOTUNE)
         if training:
-            # TODO add cropping, but crop first or rotation first?
-            affine_transform = AffineTransformation3D(
+            da_transform = AffineTransformation3D(
```
Comment (mathpluscode) [inline deepreg/dataset/loader/interface.py:128]: ideally, it'd better to have a `get_transform` function, which returns one single function (inside which you can call multiple functions), then we apply it for training only,
then depending on the configs, we can choose to apply affine transform, or ddf transform

So I suggest to
1. define a `build_transform` or `get_transform` function
2. change the config for ddf only for now, as your da config is not covering affine

what do you think @acasamitjana 

Ideally, if reviews go faster, I should change all register system before mid Dec, then the whole DA system will be easier

**Comment by YipengHu** [comment]
sorry - i'm a bit behind my plan... i will get to this tomorrow.

**Comment by acasamitjana** [comment]
Hi, sorry for the delay. I have now refactored the code using registers. I have included a gen_transform() function in the DataLoader (I'm not sure it's the best place) Please, @mathpluscode check if I'm doing it right and similar to what you are doing with backbone networks.

Now, we can append a _data_augmentation_ field in the configuration files (*.yaml). There, we can define as many preprocessing functions as we want (assuming they are implementated in _preprocess.py_ and registered in _registry.py_). For now we only have DDF since the Affine is mandatory. E.g.,
data_augmentation:
   ddf:
      param_1:
      param_2:

@YipengHu , @RemiDelaunay  can you take a look at that? 

pD: We are missing a unit test in get_preprocess_data() and gen_transforms(), but I'll leave it as the last point when everything is in palce.

**Review by YipengHu — APPROVED** [review]
@acasamitjana this looks good to me. i have not got around to figure out the failed checks, is it something to do with the register? I am approving this now, can you have a look what the best way to merge?

**Comment by mathpluscode** [comment]
@acasamitjana I suggest having a meeting regarding this PR. We can discuss a bit more about the design of registry and I can help on potential conflicts ^^ I'm free most every day.

**Comment by acasamitjana** [comment]
> @acasamitjana I suggest having a meeting regarding this PR. We can discuss a bit more about the design of registry and I can help on potential conflicts ^^ I'm free most every day.

Ok, let's do this sometime during this week. Mornings work for me at almost any time. Otherwise (after lunch/evening), let me know your availability.

File: deepreg/model/layer_util.py [hunk 2]
```diff
@@ -430,6 +430,34 @@ def random_transform_generator(
...
+    :param seed: control the randomness
+    :return:
+    """
+
+    np.random.seed(seed=seed)
+    lowres_strength = np.random.uniform(0, field_strength, (batch_size, 1, 1, 1, 3))
+    lowres_field = lowres_strength * np.random.randn(
+        batch_size, lowres_size[0], lowres_size[1], lowres_size[2], 3
+    )
+    highres_field = resize3d(lowres_field, image_size)
```
[the remark this answers is not in the document]
Comment (acasamitjana) [reply deepreg/model/layer_util.py:458]: I'd say that for small lowres_field sizes the field is sufficiently smooth. Anyway, the implementation using b-splines would be quite straightforward, even though I'm not sure how much will penalize in terms of speed.

File: deepreg/model/layer_util.py [hunk 3]
```diff
@@ -430,6 +430,34 @@ def random_transform_generator(
...
+    :param image_size:
+    :param field_strength: maximum field strength, computed as a U[0,field_strength]
+    :param lowres_size: low_resolution deformation field that will be upsampled to the original size in order to
+                        get smooth and more realistic fields.
+    :param seed: control the randomness
+    :return:
+    """
+
+    np.random.seed(seed=seed)
+    lowres_strength = np.random.uniform(0, field_strength, (batch_size, 1, 1, 1, 3))
```
[the remark this answers is not in the document]
Comment (acasamitjana) [reply deepreg/model/layer_util.py:454]: I agree on the first question, we can extend it to accept different strenghts per dimension. What do you think?
About the second, it's open to discussion. My first thought is that: the strength may depend on the input resolution (e.g: it is not the same a deformation of 3 voxels at 2mm or 1mm resolution) and the size on the number of pixels. I used these defaults with images of 1mm-resolution and  256x256 pixels.

File: deepreg/dataset/loader/interface.py [hunk 4]
```diff
@@ -9,7 +9,11 @@
 import tensorflow as tf
 
 from deepreg.dataset.loader.util import normalize_array
-from deepreg.dataset.preprocess import AffineTransformation3D, resize_inputs
+from deepreg.dataset.preprocess import (
+    AffineTransformation3D,
+    DDFTransformation3D,
```
[the remark this answers is not in the document]
Comment (acasamitjana) [reply deepreg/dataset/loader/interface.py]: Agree

**Review by mathpluscode — APPROVED** [review]

**Comment by mathpluscode** [comment]
I'm approving this PR now, but I will open a new ticket for refactoring ^^.

As you may see here https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/59c33bd7-2f0c-4e38-a126-ea647f837be6, some improvement can be done.

**Comment by mathpluscode** [comment]
Oooops! We were using affine transformation by default? Now there's no more data augmentation I guess...

**Comment by acasamitjana** [comment]
Yes, that was a default and I moved as optional. What do you think is best?

**Comment by acasamitjana** [comment]
Just saw that you added the line in all demos, thanks!

**Pull request merged**